playback/src/player.rs

```diff
@@ -1583,82 +1579,61 @@ impl PlayerInternal {
-                                    // Simplifies to:
-                                    // attack_cf * self.normalisation_peak - attack_cf * self.normalisation_integrator + self.normalisation_integrator
-                                    self.normalisation_peak = attack_cf * self.normalisation_peak
-                                        - attack_cf * self.normalisation_integrator
-                                        + self.normalisation_integrator;
-
-                                    // step 6: make-up gain applied later (volume attenuation)
-                                    // Applying the standard normalisation factor here won't work,
-                                    // because there are tracks with peaks as high as 6 dB above
-                                    // the default threshold, so that would clip.
-
-                                    // steps 7-8: conversion into level and multiplication into gain stage
-                                    *sample *= db_to_ratio(-self.normalisation_peak);
-                                }
+                                // track left/right channel
+                                let channel = self.normalisation_channel;
+                                self.normalisation_channel ^= 1;
 
-                                *sample *= volume;
+                                // step 5: smooth, decoupled peak detector
+                                self.normalisation_integrators[channel] = f64::max(
+                                    limiter_db,
+                                    release_cf * self.normalisation_integrators[channel]
+                                        + (1.0 - release_cf) * limiter_db,
+                                );
+                                self.normalisation_peaks[channel] = attack_cf
+                                    * self.normalisation_peaks[channel]
+                                    + (1.0 - attack_cf) * self.normalisation_integrators[channel];
+
+                                // steps 6-8: conversion into level and multiplication into gain
+                                // stage. Find maximum peak across both channels to couple the gain
+                                // and maintain stereo imaging.
+                                let max_peak = f64::max(
+                                    self.normalisation_peaks[0],
+                                    self.normalisation_peaks[1],
+                                );
+                                *sample *= db_to_ratio(-max_peak) * volume;
+                            }
+                        } else if self.config.normalisation_method == NormalisationMethod::Basic
+                            && (normalisation_factor < 1.0 || volume < 1.0)
+                        {
+                            for sample in data.iter_mut() {
+                                *sample *= normalisation_factor * volume;
```

Comment:
Is there a reason you move that part down? There isn't anything speaking against it, was just curious if there was a reason or if this just happened when adjusting it^^

> And wouldn't it maybe better to use a `match` rather then the `if-else` branches here? But that would only really improve handling more additions to the `NormalisationMethod` which is probably not changed anymore.

Reply:
I moved it because the default limiter is the dynamic one, so this saves at least one CPU instruction for the default case. But yeah, switching to `match` seems better. I'll update.

Reply:
Done in [56611d8](https://github.com/librespot-org/librespot/pull/1485/commits/56611d81282608cbf487d62f3c57963d6ef0fe43).